### PR TITLE
Support PWA

### PR DESCRIPTION
**Pull request by thanhnguyen2187**
Make yarr installable on desktop and mobile. Tested with:

- Vivaldi desktop
- Vivaldi mobile

<details>
    <summary>Screenshots</summary>
    
    
</details>

Also fix an edge case where HTTP Basic authentication is used and the browser won't be able to load `manifest.json` correctly.

### CODE REVIEW

**Comment by thanhnguyen2187** [comment]
This PR should address https://github.com/nkanaev/yarr/issues/48.

**Comment by nkanaev** [comment]
1. absolute links everywhere - not good. see https://github.com/nkanaev/yarr/pull/114
2. adding large amount of images is undesirable - too much fluff for too small benefit

**Comment by thanhnguyen2187** [comment]
Hi @nkanaev. I fixed the absolute links, and trimmed the icons. Can you have a look again? Thanks!

File: src/assets/index.html [hunk 1]
```diff
@@ -3,16 +3,22 @@
...
     <title>yarr!</title>
-    <link rel="stylesheet" href="./static/stylesheets/bootstrap.min.css">
-    <link rel="stylesheet" href="./static/stylesheets/app.css">
-    <link rel="icon" href="./static/graphicarts/favicon.svg" type="image/svg+xml">
-    <link rel="alternate icon" href="./static/graphicarts/favicon.png" type="image/png">
-    <link rel="manifest" href="./manifest.json" />
+    <link rel="stylesheet" href="./static/stylesheets/bootstrap.min.css" crossorigin="use-credentials">
+    <link rel="stylesheet" href="./static/stylesheets/app.css" crossorigin="use-credentials">
+    <link rel="icon" href="./static/graphicarts/favicon.svg" type="image/svg+xml" crossorigin="use-credentials">
+    <link rel="alternate icon" href="./static/graphicarts/favicon.png" type="image/png" crossorigin="use-credentials">
```
Comment (nkanaev) [inline src/assets/index.html:9]: are these required? what's the purpose of the additional tags?

File: src/assets/javascripts/sw.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+const VERSION = "v2.4"
```
Comment (nkanaev) [inline src/assets/javascripts/sw.js:1]: I'd rather not have it pinned in the source code. `makefile` should be the only source of truth for the version (if it's absolutely needed)

File: src/assets/javascripts/sw.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+const APP_STATIC_RESOURCES = [
+  "/",
+  "/static/stylesheets/bootstrap.min.css",
+  "/static/stylesheets/app.css",
+  "/static/graphicarts/favicon.svg",
+  "/static/graphicarts/favicon.png",
```
Comment (nkanaev) [inline src/assets/javascripts/sw.js:7]: are these guaranteed to work with the custom `-base` parameter set?

File: src/server/routes.go [hunk 4]
```diff
@@ -93,10 +94,24 @@ func (s *Server) handleManifest(c *router.Context) {
 				"sizes": "64x64",
 				"type":  "image/png",
 			},
+			{
+				"src":   s.BasePath + "/static/graphicarts/favicon-144.png",
```
Comment (nkanaev) [inline src/server/routes.go:98]: is png a must? wouldn't svg be sufficient? why 144 (out of curiosity)?

File: src/server/routes.go [hunk 5]
```diff
@@ -93,10 +94,24 @@ func (s *Server) handleManifest(c *router.Context) {
...
+				"src":   s.BasePath + "/static/graphicarts/favicon.svg",
+				"sizes": "any",
+				"type":  "image/svg",
+			},
 		},
 	})
 }
 
+func (s *Server) handleServiceWorker(c *router.Context) {
+	http.ServeFile(c.Out, c.Req, "src/assets/javascripts/sw.js")
```
Comment (nkanaev) [inline src/server/routes.go:112]: is it guaranteed to work for a compiled binary (which embeds assets)?

**Review by nkanaev — COMMENTED** [review]
sorry for the late reply. could you clarify few moments?

File: src/assets/index.html (hunk 1 above)
Comment (thanhnguyen2187) [reply src/assets/index.html:9]: `crossorigin="use-credentials"` is needed if we use Basic Authentication of HTTP. It is an edge case where I deploy yarr using the HTTP server's Basic Authentication (Caddy in this case). The browser is able to load `manifest.json` correctly, but would not be able to load other files specified by `manifest.json`.

File: src/assets/javascripts/sw.js (hunk 2 above)
Comment (thanhnguyen2187) [reply src/assets/javascripts/sw.js:1]: I got it. Do you have any suggestion on how should I include the version here?

File: src/server/routes.go (hunk 5 above)
Comment (thanhnguyen2187) [reply src/server/routes.go:112]: Yeah I see. Do you have any suggestion on how should I handle this service worker file?

File: src/server/routes.go (hunk 4 above)
Comment (thanhnguyen2187) [reply src/server/routes.go:98]: That's what my browser (Vivaldi or Chromium-based browsers) requires for an installable PWA. You can turn on the developer console and switch to tab "Application" and see how it goes.

image

File: src/assets/javascripts/sw.js (hunk 3 above)
Comment (thanhnguyen2187) [reply src/assets/javascripts/sw.js:7]: I did not test that case, but it does look like relative links should be preferred.

**Pull request closed without merging**